src/main/java/org/radarbase/authorizer/service/managementportal/CachedManagementPortalClient.java

```diff
@@ -0,0 +1,218 @@
+  public CachedManagementPortalClient(@Value("${mp-config-file-path}") String mpConfigFilePath)
+      throws MalformedURLException, ConfigurationException {
+    subjects = new HashSet<>();
+    projects = new HashSet<>();
+    lastFetch = Instant.MIN;
+    init(mpConfigFilePath);
+  }
+
+  public CachedManagementPortalClient(@Value("${mp-config-file-path}") String mpConfigFilePath,
+      Duration expiry) throws MalformedURLException, ConfigurationException {
+    this.expiry = expiry;
+    subjects = new HashSet<>();
+    projects = new HashSet<>();
+    lastFetch = Instant.MIN;
+    init(mpConfigFilePath);
+  }
+
+  private void init(String mpConfigFilePath) throws ConfigurationException, MalformedURLException {
+
+    this.httpClient = new OkHttpClient.Builder()
+        .connectTimeout(20, TimeUnit.SECONDS)
+        .writeTimeout(20, TimeUnit.SECONDS)
+        .readTimeout(50, TimeUnit.SECONDS)
+        .build();
+
+    this.mapper = new ObjectMapper()
+        .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+
+    this.properties = loadManagementPortalProperties(mpConfigFilePath);
+    this.oAuth2Client = new OAuth2Client.Builder()
+        .credentials(properties.getOauthClientId(), properties.getOauthClientSecret())
+        .endpoint(new URL(properties.getBaseUrl()), "/oauth/token")
+        .httpClient(httpClient)
+        .build();
+  }
+
+  private ManagementPortalProperties loadManagementPortalProperties(@NotNull String path)
+      throws ConfigurationException {
+    ManagementPortalProperties properties = ConfigHelper
+        .loadPropertiesFromFile(path, new TypeReference<ManagementPortalProperties>() {
+        });
+
+    if (properties.anyNull()) {
+      throw new ConfigurationException(
+          "Some of the required properties were not defined in " + path);
+    }
+
+    return properties;
+  }
+
+  @Override
+  public Subject getSubject(String subjectId) throws IOException, TokenException {
+    // First check if need to refresh subjects cache
+    if (isUpdateRequired()) {
+      update();
+      return this.subjects.stream()
+          .filter(subject1 -> subject1.getSubjectId().equals(subjectId))
+          .findFirst()
+          .orElse(null);
+    }
+
+    // Try to find the subject in cache if not updated
+    Optional<Subject> subject = this.subjects.stream()
+        .filter(subject1 -> subject1.getSubjectId().equals(subjectId))
+        .findFirst();
+
+    // Try to get from MP if not present in cache.
+    return subject.orElse(querySubject(subjectId));
+  }
+
+  @Override
+  public Project getProject(String projectId) throws IOException, TokenException {
+    if (isUpdateRequired()) {
+      update();
+      return this.projects.stream()
+          .filter(project1 -> project1.getProjectId().equals(projectId))
+          .findFirst()
+          .orElse(null);
+    }
+    Optional<Project> project = this.projects.stream()
+        .filter(project1 -> project1.getProjectId().equals(projectId))
```

Comment:
why not find?

Reply:
The stream doesn't have a find(predicate) function. Am i looking in the worng place?